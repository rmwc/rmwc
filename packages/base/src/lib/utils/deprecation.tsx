export type DeprecateT = {
  [oldPropName: string]: string | [string, (value: any) => void];
};

export const deprecationWarning = (message: string) => {
  //if (!import.meta.env.PROD) {
  //  console.warn(`RMWC Deprecation Warning: ${message}`);
  //}
};

export const handleDeprecations = (
  props: any,
  deprecate: DeprecateT,
  displayName: string
) => {
  props = { ...props };
  for (const oldPropName in deprecate) {
    const newProp = deprecate[oldPropName];
    let newPropName;
    let transformProp = (value: any) => value;

    if (Array.isArray(newProp)) {
      newPropName = newProp[0];
      transformProp = newProp[1];
    } else {
      newPropName = newProp;
    }

    if (props[oldPropName] !== undefined) {
      if (newPropName === '') {
        /* istanbul ignore next */
        deprecationWarning(
          `${
            displayName || ''
          } component prop '${oldPropName}' has been removed from and is no longer a valid prop.`
        );
      } else {
        props[newPropName] = transformProp(props[oldPropName]);
        let propTransformMessage = '';
        if (props[newPropName] !== props[oldPropName]) {
          propTransformMessage = ` The old value has also been converted from '${props[oldPropName]}' to '${props[newPropName]}'`;
        }

        /* istanbul ignore next */
        deprecationWarning(
          `${
            displayName || ''
          } component prop '${oldPropName}' has been replaced with '${newPropName}'. ${propTransformMessage}`
        );
      }

      delete props[oldPropName];
    }
  }
  return props;
};
