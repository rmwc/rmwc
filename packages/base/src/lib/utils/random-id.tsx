/**
 * Generates a pseudo random string for DOM ids
 * Will return 'test' in the NODE test-env so things like storyshots doesnt break.
 */
export const randomId = (prefix?: string): string => {
  const id =
    //import.meta.env.MODE === 'test'
    //  ? 'test'
    //  :
    (Math.random() + Math.random() + 1).toString(36).substring(2);
  return `${prefix}-${id}`;
};
