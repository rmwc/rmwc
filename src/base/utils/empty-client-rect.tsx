// as of mcw 14, mcw themselves has typed it as any
export const emptyClientRect: any = {
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0
};
