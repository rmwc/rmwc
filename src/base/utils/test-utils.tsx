import React from 'react';
import { act } from 'react-dom/test-utils';

export const wait = (timeout = 0) =>
  new Promise((resolve) => {
    setTimeout(resolve, timeout);
  });

export const actWait = async (timeout = 0) => {
  await act(async () => {
    await wait(timeout);
  });
};
