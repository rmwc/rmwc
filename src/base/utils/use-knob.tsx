/* istanbul ignore file */
import { useState } from 'react';
import * as knobTypes from '@storybook/addon-knobs';

// @ts-ignore
import { manager } from '@storybook/addon-knobs/dist/esm/registerKnobs';

export const useKnob = (
  knobType: keyof typeof knobTypes,
  name: string,
  defaultValue: any
) => {
  const knobFunc = knobTypes[knobType] as any;
  const [stateValue, _stateSetter] = useState(defaultValue);
  const knobValue = knobFunc(name, stateValue);

  const stateSetter = (value: any) => {
    _stateSetter(value);
    manager.knob(name, stateValue);
  };

  if (knobValue !== stateValue) {
    stateSetter(knobValue);
  }
  return [stateValue, stateSetter];
};
