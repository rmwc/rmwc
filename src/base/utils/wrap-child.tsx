import React from 'react';
import classNames from 'classnames';

export const wrapChild = (props: { [key: string]: any }) => {
  const child = React.Children.only(props.children);
  return React.cloneElement(child, {
    ...props,
    ...child.props,
    className: classNames(props.className, child.props.className),
    style: { ...child.props.style, ...props.style }
  });
};
