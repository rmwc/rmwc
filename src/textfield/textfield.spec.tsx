import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { TextField, TextFieldHelperText } from './';
import { wait } from '@rmwc/base/utils/test-utils';

describe('TextField', () => {
  it('renders', () => {
    const { asFragment } = render(
      <TextField label="test" placeholder="test" />
    );
    expect(screen.getByText('test')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('can autoFocus', () => {
    const { container } = render(
      <TextField label="test" placeholder="test" autoFocus />
    );
    expect(document.activeElement).toBe(container.querySelector('input'));
  });

  it('can have children', () => {
    const { asFragment } = render(
      <TextField placeholder="test">
        <div>Child</div>
      </TextField>
    );
    expect(screen.getByText('Child')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('can have helpText', () => {
    render(
      <div>
        <TextField helpText="textHelpText1" />
      </div>
    );
    const { container } = render(
      <div>
        <TextField
          helpText={{
            children: 'textHelpText2',
            validationMsg: true,
            persistent: true
          }}
        />
      </div>
    );

    expect(screen.getByText('textHelpText1')).toBeInTheDocument();
    expect(screen.getByText('textHelpText2')).toBeInTheDocument();

    expect(
      container.getElementsByClassName(
        'mdc-text-field-helper-text--validation-msg'
      )
    ).toHaveLength(1);
    expect(
      container.getElementsByClassName('mdc-text-field-helper-text--persistent')
    ).toHaveLength(1);
  });

  it('can have custom classnames', () => {
    const { container } = render(
      <TextField placeholder="test" className="my-custom-classname">
        <div>Child</div>
      </TextField>
    );

    expect(container.firstChild).toHaveClass('my-custom-classname');
  });

  it('can be bound', () => {
    render(
      <TextField placeholder="test" value="hello world" onChange={() => {}} />
    );
    expect(screen.getByRole('textbox')).toHaveValue('hello world');
  });

  it('can be textarea', () => {
    const { asFragment } = render(
      <TextField
        placeholder="test"
        value="hello world"
        textarea
        onChange={() => {}}
      />
    );
    expect(screen.getByText('hello world')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('can have custom classnames on input', () => {
    const { container } = render(
      <TextField className={'my-custom-classname'} />
    );
    expect(container.firstChild).toHaveClass('my-custom-classname');
  });

  it('can be invalid', () => {
    const { asFragment } = render(<TextField invalid />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('can be outlined', () => {
    const { asFragment } = render(<TextField outlined />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('can be disabled', () => {
    const { asFragment } = render(<TextField disabled />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('can be required', async () => {
    const { container } = render(
      <TextField value="" onChange={() => {}} required />
    );
    const getValid = () =>
      container.getElementsByClassName('mdc-text-field--invalid').length === 0;

    // should render valid to start
    expect(getValid()).toBe(true);

    fireEvent.focus(screen.getByRole('textbox'));
    await wait(20);
    fireEvent.blur(screen.getByRole('textbox'));
    await wait(20);

    expect(getValid()).toBe(false);
  });

  it('can be have icon', () => {
    const { asFragment } = render(<TextField icon="favorite" />);
    expect(screen.getByText('favorite')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('can be have trailingIcon', () => {
    const { asFragment } = render(<TextField trailingIcon="favorite" />);
    expect(screen.getByText('favorite')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('supports inputRef as an object reference', () => {
    const textObjectRef: any = { current: null };
    const { rerender } = render(<TextField inputRef={textObjectRef} />);
    expect(textObjectRef.current instanceof HTMLInputElement).toBeTruthy();

    const areaObjectRef: any = { current: null };
    rerender(<TextField inputRef={areaObjectRef} textarea />);
    expect(areaObjectRef.current instanceof HTMLTextAreaElement).toBeTruthy();
  });

  it('supports inputRef as a function reference', () => {
    let inputObjectRef: any;
    const objectRefFunc: any = (el: HTMLInputElement) => {
      inputObjectRef = el;
    };
    render(<TextField inputRef={objectRefFunc} />);
    expect(inputObjectRef instanceof HTMLInputElement).toBeTruthy();
  });

  it('label floats on dynamic change', async () => {
    const { container } = render(
      <TextField label="test" value="" onChange={() => {}} />
    );
    expect(
      container.getElementsByClassName('mdc-floating-label--float-above')
    ).toHaveLength(0);
    userEvent.type(screen.getByRole('textbox'), 'foo');
    await wait(100);
    expect(
      container.getElementsByClassName('mdc-floating-label--float-above')
    ).toHaveLength(1);
  });

  it('can have a prefix', () => {
    render(<TextField prefix="USD" />);
    expect(screen.getByText('USD')).toBeInTheDocument();
  });

  it('can have a suffix', () => {
    render(<TextField suffix="USD" />);
    expect(screen.getByText('USD')).toBeInTheDocument();
  });
});

describe('TextFieldHelperText', () => {
  it('renders', () => {
    const { asFragment } = render(
      <TextFieldHelperText>Hello</TextFieldHelperText>
    );
    expect(screen.getByText('Hello')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
