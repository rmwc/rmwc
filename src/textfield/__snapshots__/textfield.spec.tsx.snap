// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField can be disabled 1`] = `
<DocumentFragment>
  <label
    aria-labelledby="textfield-test-label"
    class="mdc-text-field mdc-text-field--upgraded mdc-text-field--filled mdc-text-field--disabled mdc-text-field--no-label"
  >
    <span
      class="mdc-text-field__ripple"
    />
    <input
      class="mdc-text-field__input"
      disabled=""
    />
    <span
      class="mdc-line-ripple mdc-line-ripple--deactivating"
      style="transform-origin: 0px center;"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField can be have icon 1`] = `
<DocumentFragment>
  <label
    aria-labelledby="textfield-test-label"
    class="mdc-text-field mdc-text-field--upgraded mdc-text-field--filled mdc-text-field--with-leading-icon mdc-text-field--no-label"
  >
    <i
      class="rmwc-icon rmwc-icon--ligature material-icons mdc-text-field__icon mdc-text-field__icon--leading"
    >
      favorite
    </i>
    <span
      class="mdc-text-field__ripple"
    />
    <input
      class="mdc-text-field__input"
    />
    <span
      class="mdc-line-ripple mdc-line-ripple--deactivating"
      style="transform-origin: 0px center;"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField can be have trailingIcon 1`] = `
<DocumentFragment>
  <label
    aria-labelledby="textfield-test-label"
    class="mdc-text-field mdc-text-field--upgraded mdc-text-field--filled mdc-text-field--with-trailing-icon mdc-text-field--no-label"
  >
    <span
      class="mdc-text-field__ripple"
    />
    <input
      class="mdc-text-field__input"
    />
    <i
      class="rmwc-icon rmwc-icon--ligature material-icons mdc-text-field__icon mdc-text-field__icon--trailing"
      tabindex="0"
    >
      favorite
    </i>
    <span
      class="mdc-line-ripple mdc-line-ripple--deactivating"
      style="transform-origin: 0px center;"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField can be invalid 1`] = `
<DocumentFragment>
  <label
    aria-labelledby="textfield-test-label"
    class="mdc-text-field mdc-text-field--upgraded mdc-text-field--filled mdc-text-field--invalid mdc-text-field--no-label"
  >
    <span
      class="mdc-text-field__ripple"
    />
    <input
      class="mdc-text-field__input"
    />
    <span
      class="mdc-line-ripple mdc-line-ripple--deactivating"
      style="transform-origin: 0px center;"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField can be outlined 1`] = `
<DocumentFragment>
  <label
    aria-labelledby="textfield-test-label"
    class="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined mdc-text-field--no-label"
  >
    <span
      class="mdc-text-field__ripple"
    />
    <input
      class="mdc-text-field__input"
    />
    <div
      class="mdc-notched-outline mdc-notched-outline--no-label"
    >
      <div
        class="mdc-notched-outline__leading"
      />
      <div
        class="mdc-notched-outline__notch"
      />
      <div
        class="mdc-notched-outline__trailing"
      />
    </div>
  </label>
</DocumentFragment>
`;

exports[`TextField can be textarea 1`] = `
<DocumentFragment>
  <label
    aria-labelledby="textfield-test-label"
    class="mdc-text-field mdc-text-field--upgraded mdc-text-field--filled mdc-text-field--textarea mdc-text-field--no-label"
  >
    <span
      class="mdc-text-field__ripple"
    />
    <textarea
      aria-invalid="false"
      class="mdc-text-field__input hf-validated hf-valid hf-in-range"
      placeholder="test"
    >
      hello world
    </textarea>
    <span
      class="mdc-line-ripple mdc-line-ripple--deactivating"
      style="transform-origin: 0px center;"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField can have children 1`] = `
<DocumentFragment>
  <label
    aria-labelledby="textfield-test-label"
    class="mdc-text-field mdc-text-field--upgraded mdc-text-field--filled mdc-text-field--no-label"
  >
    <div>
      Child
    </div>
    <span
      class="mdc-text-field__ripple"
    />
    <input
      class="mdc-text-field__input"
      placeholder="test"
    />
    <span
      class="mdc-line-ripple mdc-line-ripple--deactivating"
      style="transform-origin: 0px center;"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField renders 1`] = `
<DocumentFragment>
  <label
    aria-labelledby="textfield-test-label"
    class="mdc-text-field mdc-text-field--upgraded mdc-text-field--filled"
  >
    <span
      class="mdc-text-field__ripple"
    />
    <input
      class="mdc-text-field__input"
      placeholder="test"
    />
    <span
      class="mdc-floating-label"
      id="textfield-test-label"
    >
      test
    </span>
    <span
      class="mdc-line-ripple mdc-line-ripple--deactivating"
      style="transform-origin: 0px center;"
    />
  </label>
</DocumentFragment>
`;

exports[`TextFieldHelperText renders 1`] = `
<DocumentFragment>
  <div
    class="mdc-text-field-helper-text"
  >
    Hello
  </div>
</DocumentFragment>
`;
